### ISSUE TITLE
[jvm-packages] xgboost4j - Labeled Point - Data point weights

### ISSUE THREAD

**Issue by RhariharanExpedia** [dmlc/xgboost#2122]
There is a tempting notion that one day distributed XGBoost will allow _**data point weights**_. This is currently a feature in the single-machine API for Python.  

In [this file](https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/LabeledPoint.java), LabeledPoint contains a field for "weight," which seems to suggest that someone was thinking about having this feature in the distributed version at some point.

Is adding "data point weight" to distributed XGBoost already in progress somewhere?  I would be interested in collaborating with you if you are already working on this.

**Comment by CodingCat** [comment]
Feel free to start, I haven't started anything related to this

**Comment by RhariharanExpedia** [comment]
After spending some time reading through the code I'm trying to identify the best route to implement this change. 

The options I see are generally:

- Add an additional optional parameter for many of the functions so that weights can travel through the various layers separately. Is there a risk that through various transitions, weights no longer match up to their LabeledPoint?

- Add this feature through trainWithDataFrame (in spark/xgboost.scala, line 164). Since the client could easily add a column "weights" to the Dataframe, it would make this change more straightforward and reduce impact on clients who don't need weights. This would also require a somewhat parallel path to generating the DMatrix objects. While they have a _weight_ property, the code in between uses MLLabeledPoint, so changes would be needed there.

- Update the code using RDD's to translate data into a Dataset and use that internally to pass data / iterators around. So instead of the trainWithDataFrame code ultimately converting data to RDD's on its way to the XGBoost C++ library, the opposite would be true -- the RDD method would convert the data to a DataFrame for passing around internally. 

Related Questions:
- What are your plans around reducing RDD support in favor of Datasets? 
- Should spark/XGBoost.scala line 187 call .train rather than .fit? Fit seems to call the spark.ml method, but train calls the XGBoost code. Or maybe I missed something?

Your thoughts about this are welcome. I want to make sure I'm moving the code forward consistent with other efforts, but also want to reduce the amount of code to rewrite. 

Thoughts?

**Comment by CodingCat** [comment]
> What are your plans around reducing RDD support in favor of Datasets?

Basically, I will try to avoid involving Dataset until the community has shown that it is the next promoted thing. Until so far, Spark SQL APIs still works around DataFrame, the same to ML. 

> Should spark/XGBoost.scala line 187 call .train rather than .fit? Fit seems to call the spark.ml method, but train calls the XGBoost code. Or maybe I missed something?

No, should call fit() to go into XGBoostEstimator

**Comment by CodingCat** [comment]
We may discard the support weight with RDD-based API, we can do the following items

* allow the user optionally pass in a weight column

* modify this part to directly convert to XGBoost LabeledPoint if weight column presents https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostEstimator.scala#L115

* add a `trainWithRDDInternal` to take XGBoost LabeledPoint, the modified part  in the above item shall directly call this `trainWithRDDInternal` (should extract the common part in the current implementation)

**Comment by CodingCat** [comment]
close for no update

**Comment by RhariharanExpedia** [comment]
I moved on to something else for a bit, but I am still working on this.

**Comment by ebernhardson** [comment]
Rather than extracting  commonality just trainWithRDD, what do you think about moving all that code to trainWithRDDInternal and changing the signature to Either[RDD[LabeledPoint], RDD[MLLabelefPoint]], then handling the differences with pattern matching? This would simplify a bit, as there are several methods, such as buildDistributedBoostèrs, that would also require a refactor. This might be a little messier than a full refactoring though.

**Comment by CodingCat** [comment]
hmmm...as a downstream project, I do not want to put effort to handle RDD stuffs, essentially RDD-based solutions are fading out in Spark's own submodules... so...instead of introducing Either, etc. to take care about old RDD...I prefer to just leave it as it is, on some version, mark it as deprecated....

**Comment by CodingCat** [comment]
@ebernhardson , if you are working on this, feel free to open it

**Comment by ebernhardson** [comment]
i might find some time to work on this, but still trying to figure out what the right approach will be. All training interfaces currently flow to trainWithRDD eventually, it's where the magic happens. If the goal is to deprecate is seems that will all need to move at some point, but for now it's what we have.

What if trainWithRDD implementation basically became:

  val convertedTraininedData = trainingData.mapPartitions(DataUtils.fromSparkPointsToXGBoostPoints)
  trainWithRDDInternal(convertedTrainingData, ...)

the current implementation would move to trainWithRDDInternal with the new signature RDD[LabeledPoint], other signatures for private methods would also change signatures. The fromDenseToSparseLabeledPoints would be rewritten to use the XGB LabeledPoint. I'm thinking most obvious might be to add a toSparse(float missing) method to xgb's LabeledPoint class, which would either do nothing when already sparse, or convert itself from dense to sparse by removing. the fromDenseToSparseLabeledPoints method then simply iterates the list when missing is not Float.isNaN and calls point.toSparse(missing)

Seem reasonable?

**Comment by RhariharanExpedia** [comment]
@ebernhardson - Since the DMatrix objects that ultimately get sent to the C++ code have a weight concept, it seemed like things should work like the standalone version in the layers below JNI call. My approach was to have parallel paths below trainWithDataFrame (one using XGB.LabeledPoint, the other using the original ml.LabeledPoint), but I really didn't like the duplication of code that approach entails.

If you'd be up for a discussion about different options, I can set something up.

**Comment by BigPotato** [comment]
hi @CodingCat , 
my colleagues are working with c++ xgboost. their training data is label with weight like this:
(label:weight  features ...)
0:1  10:0.1  11:0.2  12:0.3
1:2  10:0.5  11:0.8  12:0.1

now i am trying xgboost4j, but i can not load the training data in the label-weight format.
if i flat the traing data to be this:
0  10:0.1  11:0.2  12:0.3
1  10:0.5  11:0.8  12:0.1
1  10:0.5  11:0.8  12:0.1

then train the model with  xgboost4j，is it equivalent to training by c++ version with label-weight data?

Thanks a lot !
